docs/decisions/0002-enterprise-catalog-ida.rst

```diff
@@ -0,0 +1,64 @@
+
+What is Enterprise Catalog?
+___________________________
+
+Enterprise catalogs are essentially saved elasticsearch queries attached to an Enterprise Customer. There can be many catalogs configured for a given customer, which enables different sets of learners from the same Enterprise to be given access to different sets of courses, and allows different discounts to be configured for different sets of courses.
+
+We expose two apis that are frequently used by various enterprise features: one to retrieve all of the content metadata that would be returned by the configured catalog query, and one to determine if a given course id or program id is part of the set of content returned by the catalog query. These apis exist both as python apis and REST apis that can be hit from within our edX ecosystem and externally from customers. They are wrappers around the search/all endpoint in the discovery service. Although there is caching in place, the calls to discovery are implemented synchronously, and we've found that half the time requests do not hit the cache, and thus the most of the time is spent waiting on the discovery service to return results.
+
+Who’s Calling What now?
+_______________________
+
+Here is a diagram showing a simplified view of the calls to these apis. The yellow boxes represent the features or user workflows that trigger calls to the various apis. The blue clouds represent REST apis, and the purple rounded boxes represent python code apis. Solid lines are synchronous api calls and dotted lines are code apis (not hitting the network).
+
+.. image:: ../images/enterprise-catalog-api-calls.png
+
+Decisions
+=========
+
+In order to address these performance issues and to more effectively scale and build new features related to enterprise catalogs in the future, we will create a new IDA (Independently Deployable Application) to replace the existing enteprise catalog functionality.
+This IDA will support the existing API contracts established in the edx-enterprise API, but will be a new implementation separate from what we do today.
+
+The microservice will be supported by the following relational data model:
+
+.. image:: ../images/enterprise-catalog-ida-uml-diagram.png
+
+The microservice will expose the following APIs for use:
+
+* CRUD Operations for Enterprise Catalog
+* is content id in catalog: Given a course, course run, program, or other content id, and an enterprise catalog, returns true if the content id is contained within the catalog's current results.
+* get_catalog_metadata: Given an enterprise catalog, return the full content metadata results for content currently in the catalog.
+* refresh_catalog_metadata: Given an enterprise catalog, kick off a task to get the freshest content metadata for that catalog.
```

Comment:
This might just be me carrying baggage from discovery land, but this is a little scary. Data freshness isn't a bad thing, but we've dealt with lots of pain from `refresh_course_metadata` and I'm worried about this becoming something like that.
At the most trivial level, I'd suggest distancing ourselves from the RCM name

Reply:
happy to change names. i think a lot of the issues with discovery (from my understanding) stem from the fact that it is managing bi-directional reads and writes. the discovery service should never be aware of this service, compared to discovery needing to be concerned with lms/ecommerce/drupal. also, though we may have to support search functionality in the future, for now we are not dealing with elasticsearch, and when we do hopefully we can use the latest version. my understanding is that a lot of the time in the search/all endpoint is spent with elasticsearch. 
i think your concerns are legitimate though. ultimately, i'm betting on us being able to have more control over what happens, because this is owned by just us, and not another team. having that control will enable us to change direction much more easily if something isn't working the way we want.

Reply:
👍 I'm on board with that. As long as we're aware of that risk and have a plan of avoiding it going forward (which I think we do), then things should be fine